File: Easy/Prob_733/flood_fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */
var floodFill = function (image, sr, sc, newColor) {
  let height = image.length,
    width = image[0].length,
    pcolor = image[sr][sc];

  const dfs = function (sr, sc) {
    if (
      sr < 0 ||
      sc < 0 ||
      sr > height - 1 ||
      sc > width - 1 ||
      image[sr][sc] === newColor ||
      image[sr][sc] != pcolor
    )
      return;
    image[sr][sc] = newColor;
    dfs(sr, sc + 1);
    dfs(sr, sc - 1);
    dfs(sr + 1, sc);
    dfs(sr - 1, sc);
  };
  dfs(sr, sc);

  return image;
};

File: Easy/Prob_733/flood_fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        rows = len(image)
        cols = len(image[0])
        cur_color = image[sr][sc]
        
        def dfs(sr, sc):
            nonlocal rows, cols, newColor, image
            
            if sr < 0 or sc < 0 or sr>rows-1 or sc>cols-1 or image[sr][sc]==newColor or image[sr][sc]!=cur_color:
                return
            
            image[sr][sc] = newColor
            
            dfs(sr+1,sc)
            dfs(sr-1,sc)
            dfs(sr,sc+1)
            dfs(sr,sc-1)
        
        dfs(sr, sc)
        
        return image